tutorials/grails3-send-push-notification-using-firebase-cloud-messaging

```diff
@@ -0,0 +1,196 @@
+
+1. Open terminal or command line then go to your projects folder.
+2. Type this command.
+   ```
+   grails create-app grails-fcm
+   ```
+3. Go to the newly created Grails 3 application folder.
+   ```
+   cd grails-fcm
+   ```
+4. Enter Grails 3 interactive console by type this command.
+   ```
+   grails
+   ```
+5. Now the you are in Grails interactive console.
+   ```
+   | Resolving Dependencies. Please wait...
+   
+   CONFIGURE SUCCESSFUL
+
+   | Enter a command name to run. Use TAB for completion:
+   grails>
+   ```
+6. Run the Grails web application to make sure everything working properly by type this command in the Grails interactive console.
+   ```
+   run-app
+   ```
+## Add Grails Datastore Rest Client Dependency
+
+We will accessing Firebase using REST API, for that we need to add Grails Datastore Client Rest Dependency in our Grails 3 application. To do that, open and edit build.gradle then add this dependency inside dependencies section. There are two dependencies, one inside buildscript and the other out side buildscripts.
+```
+dependencies {
+    ...
+    compile "org.hibernate:hibernate-core:5.1.2.Final"
+    compile "org.hibernate:hibernate-ehcache:5.1.2.Final"
+    compile 'org.grails:grails-datastore-rest-client:5.0.0.RC2'
+    console "org.grails:grails-console"
+    profile "org.grails.profiles:web"
+    runtime "com.bertramlabs.plugins:asset-pipeline-grails:2.11.6"
+    ...
+}
+```
+Compile the Grails 3 application by type this command in the Grails interactive console.
+```
+compile
+```
+## Create Grails Controller and View for Sending Push Notification
+
+Do this following tasks:
+
+1. Create controller by type this command.
+   ```
+   create-controller Fcm
+   ```
+2. Add import to the file "grails-app/controllers/grails/fcm/FcmController.groovy".
+   ```
+   import grails.plugins.rest.client.RestBuilder
+   ```
+3. Create a method to FcmController, so it will looks like this.
+   ```
+    package grails.fcm
+
+    import grails.plugins.rest.client.RestBuilder
+
+    class FcmController {
+
+    def index() { }
+
+    def sendPushNotification() {
+     def regid = params.regid
+     def title = params.title
+     def body = params.body
+
+     def rest = new RestBuilder(connectTimeout:1000, readTimeout:20000)
+     def resp = rest.post("https://fcm.googleapis.com/fcm/send") {
+       header 'Content-Type', 'application/json'
+       header 'Authorization', 'key=AIza*****'
+       json {
+         notification = {
+           title = title
+           body = body
+           sound = "default"
+           click_action = "FCM_PLUGIN_ACTIVITY"
+           icon = "fcm_push_icon"
+         }
+         to = regid
+       }
+     }
+
+     flash.message = "Notification sent"
+     redirect action: "index"
+    }
+   }
+   ```
+    Your previous written FCM legacy server key write on this line include with 'key='.
```

Comment:
Enter the Legacy Server key you saved from your project dashboard in the line that begins with 'key='